habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/util/DockerfileUtil.java

```diff
@@ -0,0 +1,104 @@
+    public static final String HABUSHU_COMMENT_END = " - HABUSHU GENERATED CODE (END)";
+    public static final String REPLACE_WITH_SINGLE_REPO_PROJECT_DIR = "REPLACE_WITH_SINGLE_REPO_PROJECT_DIR";
+    public static final String ANCHOR_DIRECTORY = "ANCHOR_DIRECTORY";
+
+    /**
+     * Create the container stage content with given template, anchor directory and the single module's base directory
+     * @param template container stage template name
+     * @param anchorDirectory the anchor directory
+     * @param moduleBaseDir the module base directory
+     * @return
+     */
+    public static String getContainerStageContentFrom(String template, String anchorDirectory, String moduleBaseDir) {
+        StringBuilder content = new StringBuilder();
+        InputStream inputStream = DockerfileUtil.class.getClassLoader().getResourceAsStream(template);
+
+        try (InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
+             BufferedReader buffer = new BufferedReader(inputStreamReader)) {
+            String line = buffer.readLine();
+
+            while (line != null) {
+                line = line.strip();
+                if (line.contains(ANCHOR_DIRECTORY)) {
+                    line = line.replaceAll(ANCHOR_DIRECTORY, anchorDirectory);
+                }
+                if (line.contains(REPLACE_WITH_SINGLE_REPO_PROJECT_DIR)) {
+                    line = line.replaceAll(REPLACE_WITH_SINGLE_REPO_PROJECT_DIR, moduleBaseDir);
+                }
+                content.append(line).append("\n");
+                line = buffer.readLine();
+            }
+            inputStream.close();
+        } catch (IOException e) {
+            throw new HabushuException("Could not read from file.", e);
+        }
+        return content.toString();
+    }
+
+    /**
+     * Update the Dockerfile with container stage logic
+     * @param dockerFile the Dockerfile to be updated
+     * @param builderStageContent the builder stage content
+     * @param finalStageContent the final stage content
+     * @return
+     */
+    public static String updateDockerfileWithContainerStageLogic(File dockerFile, String builderStageContent, String finalStageContent) {
+        StringBuilder content = new StringBuilder();
+        boolean skipLine = false;
+        try (BufferedReader buffer = new BufferedReader(new FileReader(dockerFile))) {
+            String line = buffer.readLine();
+
+            while (line != null) {
+                line = line.strip();
+
+                // start skipping the line if reads HABUSHU_COMMENT_START
+                if (!skipLine && line.contains(HABUSHU_COMMENT_START)) {
+                    skipLine = true;
+                }
+
+                // end skipping the line when reads HABUSHU_COMMENT_END
+                if (skipLine && line.contains(HABUSHU_COMMENT_END)) {
+                    skipLine = false;
+                }
```

Comment:
Q: So based on this logic if we run into ` - HABUSHU GENERATED CODE (DO NOT MODIFY)` in the `Dockerfile` then we remove it by using `skipLine`? However, then the new content will never get added as `skipLine` is set to true

Reply:
For example, it seems to me like an input `Dockerfile` like this:
```
# syntax=docker/dockerfile:1
FROM scratch
ADD hello /

#HABUSHU_BUILDER_STAGE - HABUSHU GENERATED CODE (DO NOT MODIFY)
FROM registry.access.redhat.com/ubi9/python-311:latest AS habushu_builder
WORKDIR /work-dir
COPY ./target/containerize-support/example-monorepo ./containerize-support/example-monorepo

# Poetry and supporting plugin installations
RUN pip install poetry && \
    poetry self add poetry-monorepo-dependency-plugin && \
    poetry self add poetry-plugin-bundle

USER root
WORKDIR /work-dir/containerize-support/example-monorepo/extensions/extensions-python-dep-X
# ensure Poetry's cache directory is propertly set
ENV POETRY_CACHE_DIR="/.cache/pypoetry"
# instruct Docker to persistently store the container's Poetry cache while
# resolving dependencies during the lock process of the venv-specifier and
# building/exporting the virtual environment of the venv-specifier to /venv
RUN --mount=type=cache,target=/.cache/pypoetry/ \
    sed -i 's|develop[[:blank:]]*=[[:blank:]]*true|develop = false|g' pyproject.toml && \
    poetry lock && \
    poetry bundle venv /venv
#HABUSHU_BUILDER_STAGE - HABUSHU GENERATED CODE (END)

#HABUSHU_FINAL_STAGE - HABUSHU GENERATED CODE (DO NOT MODIFY)
# copy the pre-built venv from the builder and
COPY --from=habushu_builder /venv /venv
# configure the container to use the venv
ENV PATH="/venv/bin:$PATH"
#HABUSHU_FINAL_STAGE - HABUSHU GENERATED CODE (END)

CMD ["/hello"]
```
Would result in the following output:
```
# syntax=docker/dockerfile:1
FROM scratch
ADD hello /

CMD ["/hello"]
```

Reply:
no, the expected BUILDER/FINAL stage content will still be added since the follow on logic will add the new content if it finds the #HABUSHU_BUILDER/FINAL_STAGE tag and this is part of the final test scenario too.